flink-table/flink-table-planner/src/main/java/org/apache/flink/table/operations/AlgebraicOperationFactory.java

```diff
@@ -0,0 +1,118 @@
+import org.apache.flink.table.plan.logical.Union;
+
+import java.util.stream.IntStream;
+
+import static org.apache.flink.table.operations.AlgebraicOperationFactory.AlgebraicTableOperationType.UNION;
+
+/**
+ * Utility class for creating a valid algebraic operation such as {@link Minus}, {@link Intersect} or {@link Union}.
+ */
+@Internal
+public class AlgebraicOperationFactory {
+
+	private final boolean isStreaming;
+
+	/**
+	 * Describes what kind of operation should be created.
+	 */
+	public enum AlgebraicTableOperationType {
+		INTERSECT,
+		MINUS,
+		UNION
+	}
+
+	public AlgebraicOperationFactory(boolean isStreaming) {
+		this.isStreaming = isStreaming;
+	}
+
+	/**
+	 * Creates a valid algebraic operation.
+	 *
+	 * @param type type of operation to create
+	 * @param left first relational operation of the operation
+	 * @param right second relational operation of the operation
+	 * @param all flag defining how duplicates should be handled
+	 * @return creates a valid algebraic operation
+	 */
+	public TableOperation create(
+			AlgebraicTableOperationType type,
+			TableOperation left,
+			TableOperation right,
+			boolean all) {
+		LogicalNode leftNode = (LogicalNode) left;
+		LogicalNode rightNode = (LogicalNode) right;
+		failIfStreaming(type, all);
+		validateAlgebraicOperation(type, leftNode, right);
+		switch (type) {
+			case INTERSECT:
+				return new Intersect(leftNode, rightNode, all);
+			case MINUS:
+				return new Minus(leftNode, rightNode, all);
+			case UNION:
+				return new Union(leftNode, rightNode, all);
+		}
+		throw new TableException("Unknown algebraic operation type: " + type);
+	}
+
+	private void validateAlgebraicOperation(
+			AlgebraicTableOperationType operationType,
+			TableOperation left,
+			TableOperation right) {
+		TableSchema leftSchema = left.getTableSchema();
+		int leftFieldCount = leftSchema.getFieldCount();
+		TableSchema rightSchema = right.getTableSchema();
+		int rightFieldCount = rightSchema.getFieldCount();
+
+		if (leftFieldCount != rightFieldCount) {
+			throw new ValidationException(String.format("%s two table of different column sizes:" +
+				" %d and %d", operationType, leftFieldCount, rightFieldCount));
+		}
+
+		TypeInformation<?>[] leftFieldTypes = leftSchema.getFieldTypes();
+		TypeInformation<?>[] rightFieldTypes = rightSchema.getFieldTypes();
+		boolean sameSchema = IntStream.range(0, leftFieldCount)
+			.allMatch(idx -> leftFieldTypes[idx].equals(rightFieldTypes[idx]));
+
+		if (!sameSchema) {
+			throw new ValidationException(String.format("%s two table of different schema:" +
+				" %s and" +
+				" %s", operationType, leftSchema, rightSchema));
+		}
+	}
+
+	private void failIfStreaming(AlgebraicTableOperationType type, boolean all) {
+		boolean shouldFailInCaseOfStreaming = !all || type != UNION;
+
+		if (isStreaming && shouldFailInCaseOfStreaming) {
+			throw new ValidationException(String.format("%s on stream tables is currently not supported.", type));
```

Comment:
"The `type.toLowerCase` operation on unbounded tables is currently not supported."